### PR TITLE
Make inspector work parallel

### PR DESCRIPTION
**Pull request by vldm**

### CODE REVIEW

File: evm-utils/evm-state/src/storage/inspectors.rs [hunk 1]
```diff
@@ -1,136 +1,203 @@
...
+pub trait DataInspector<K, V> {
+    fn inspect_data(&self, key: K, value: V) -> Result<()>;
 }
 
-pub mod memorizer {
+pub trait TrieDataInsectorRaw {
+    fn inspect_data_raw<Data: AsRef<[u8]>>(&self, key: Vec<u8>, value: Data) -> Result<()>;
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd)]
```
Comment (hrls) [inline evm-utils/evm-state/src/storage/inspectors.rs:24]: `Eq`, `Ord`, `PartialOrd`, and even `PartialEq` instances are unnecessary.

File: evm-utils/evm-state/src/storage/inspectors.rs [hunk 2]
```diff
@@ -1,136 +1,203 @@
...
+pub trait TrieInspector {
+    fn inspect_node<Data: AsRef<[u8]>>(&self, trie_key: H256, node: Data) -> Result<bool>;
+}
+pub trait DataInspector<K, V> {
+    fn inspect_data(&self, key: K, value: V) -> Result<()>;
 }
 
-pub mod memorizer {
+pub trait TrieDataInsectorRaw {
+    fn inspect_data_raw<Data: AsRef<[u8]>>(&self, key: Vec<u8>, value: Data) -> Result<()>;
```
Comment (hrls) [inline evm-utils/evm-state/src/storage/inspectors.rs:21]: Probably key should be also typed as `AsRef<[u8]>`

File: evm-utils/evm-state/src/storage/inspectors.rs [hunk 3]
```diff
@@ -1,136 +1,203 @@
...
-
-    // TODO: pass storage ref into `inspect_*` method
-    // type Storage;
-    // fn inspect_raw<Data: AsRef<[u8]>>(
-    //     &mut self,
-    //     _: &Self::Storage,
-    //     (_, _): (H256, &Data),
-    // ) -> Result<bool>;
-    // fn inspect_typed(&mut self, _: &Self::Storage, _: &T) -> Result<()>;
+use std::sync::Arc;
```
Comment (hrls) [inline evm-utils/evm-state/src/storage/inspectors.rs]: Group them all

File: evm-utils/evm-state/src/storage/inspectors.rs [hunk 4]
```diff
@@ -1,136 +1,203 @@
...
+            Self: Sized;
+    }
+
+    impl TryFromSlice for H256 {
+        fn try_from_slice(slice: &[u8]) -> Result<Self>
+        where
+            Self: Sized,
+        {
+            if slice.len() != 32 {
+                bail!("Cannot get H256 from slice len:{}", slice.len())
```
Comment (hrls) [inline evm-utils/evm-state/src/storage/inspectors.rs]: `anyhow::ensure`

File: evm-utils/evm-state/src/storage/inspectors.rs [hunk 5]
```diff
@@ -1,136 +1,203 @@
...
+        pub node_size: AtomicUsize,
     }
 
-    impl Inspector<U256> for StoragesKeysCollector {
-        fn inspect_raw<Data: AsRef<[u8]>>(&mut self, key: H256, data: &Data) -> Result<bool> {
+    impl TrieInspector for TrieCollector {
+        fn inspect_node<Data: AsRef<[u8]>>(&self, key: H256, data: Data) -> Result<bool> {
             let is_new_key = self.trie_keys.insert(key);
+            self.node_size
+                .fetch_add(data.as_ref().len(), Ordering::AcqRel);
```
Comment (hrls) [inline evm-utils/evm-state/src/storage/inspectors.rs]: Actual order isnt matter IMO. We can use `Ordering::Relaxed`.

File: evm-utils/evm-state/src/storage/mod.rs [hunk 6]
```diff
@@ -288,3 +288,90 @@ pub fn default_db_opts() -> Options {
...
+            let db = self.db.borrow();
+
+            let trie_nodes = self.trie_nodes.as_ref();
+            // Cleanup unused trie keys in default column family
+            {
+                let mut batch = rocksdb::WriteBatch::default();
+
+                for (key, _data) in db.iterator(rocksdb::IteratorMode::Start) {
+                    let bytes = <&[u8; 32]>::try_from(key.as_ref())?;
+                    let key = H256::from(bytes);
```
Comment (hrls) [inline evm-utils/evm-state/src/storage/mod.rs]: `H256::try_from_slice`

File: evm-utils/evm-state/src/storage/walker.rs [hunk 7]
```diff
@@ -1,37 +1,55 @@
...
-    I: Inspector<T>,
+    DB: Borrow<rocksdb::DB> + Sync + Send,
+    TI: TrieInspector + Sync + Send,
+    DI: TrieDataInsectorRaw + Sync + Send,
 {
-    pub fn traverse(&mut self, hash: H256) -> Result<()> {
+    pub fn traverse(&self, hash: H256) -> Result<()> {
+        self.traverse_inner(Default::default(), hash)
+    }
+    pub fn traverse_inner(&self, nimble: NibbleVec, hash: H256) -> Result<()> {
```
Comment (hrls) [inline evm-utils/evm-state/src/storage/walker.rs]: typo, `nimble` => `nibble`

File: evm-utils/evm-state/src/storage/walker.rs [hunk 8]
```diff
@@ -40,53 +58,69 @@ where
...
-                    trace!("rlp: {:?}", rlp);
-                    let t = T::decode(&rlp)?;
-                    // TODO: debug T
-                    self.inspector.inspect_typed(&t)?;
-                }
-                for value in values {
-                    self.process_value(value)?;
+                // lack of copy on result, forces setting array manually
+                let mut values_result = [
+                    None, None, None, None, None, None, None, None, None, None, None, None, None,
```
Comment (hrls) [inline evm-utils/evm-state/src/storage/walker.rs:91]: `[None; 16]`

File: evm-utils/evm-state/src/storage/walker.rs [hunk 9]
```diff
@@ -40,53 +58,69 @@ where
...
+                    None, None, None,
+                ];
+                let result = rayon::scope(|s| {
+                    for (id, (value, result)) in
```
Comment (hrls) [inline evm-utils/evm-state/src/storage/walker.rs]: `id` is unclear. This is actually nibble? IMO `((nibble, value), result)` is more clever way to do the same.

File: ledger-tool/src/evm_state.rs [hunk 10]
```diff
@@ -94,24 +102,35 @@ pub fn process_evm_state_command(ledger_path: &Path, matches: &ArgMatches<'_>) -
...
+                let storages_walker = Walker::new_raw(db, trie_collector.clone(), NoopInspector);
+                for storage_root in accounts_state_walker
+                    .data_inspector
+                    .inner
+                    .storage_roots
+                    .iter()
+                {
+                    storages_walker.traverse(*storage_root)?;
+                }
+                info!("Total trie nodes info:");
```
Comment (hrls) [inline ledger-tool/src/evm_state.rs]: Header should be part of `summarize()` call.

**Review by hrls — APPROVED** [review]

File: evm-utils/evm-state/src/storage/walker.rs (hunk 8 above)
Comment (vldm) [reply evm-utils/evm-state/src/storage/walker.rs:91]: // lack of copy on result, forces setting array manually

File: evm-utils/evm-state/src/storage/mod.rs (hunk 6 above)
Comment (vldm) [reply evm-utils/evm-state/src/storage/mod.rs]: there is no try_from_slice in H256 afaik

File: ledger-tool/src/evm_state.rs (hunk 10 above)
Comment (vldm) [reply ledger-tool/src/evm_state.rs]: `TrieCollector` is generic enough, and don't have info about type of trie that it collect.

File: evm-utils/evm-state/src/storage/walker.rs (hunk 8 above)
Comment (hrls) [reply evm-utils/evm-state/src/storage/walker.rs:91]: Damn https://github.com/rust-lang/rust/issues/49147

File: evm-utils/evm-state/src/storage/mod.rs (hunk 6 above)
Comment (hrls) [reply evm-utils/evm-state/src/storage/mod.rs]: Already implemented. In this PR.
=)
https://github.com/velas/velas-chain/blob/b0ab845c8e47fd997e955886ae9c6df0a9d8e5c0/evm-utils/evm-state/src/storage/inspectors.rs#L79

File: ledger-tool/src/evm_state.rs (hunk 10 above)
Comment (hrls) [reply ledger-tool/src/evm_state.rs]: This info line doesn't contains any type specific info.

Comment (vldm) [reply ledger-tool/src/evm_state.rs]: But previous contain.

Comment (hrls) [reply ledger-tool/src/evm_state.rs]: `fn summarize(&self) where K: Debug` or something like this.

**Pull request merged**